src/common/store/handshake.ts

```diff
@@ -0,0 +1,152 @@
+import { createAction, createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit';
+import { Event } from 'electron';
+import { AnyAction } from 'redux';
+import { eventChannel, SagaIterator } from 'redux-saga';
+import { all, call, put, select, take, takeEvery, takeLatest } from 'redux-saga/effects';
+
+import {
+  createHandshakeKeyPair,
+  decryptJson,
+  isEncryptedAction,
+  isReduxAction
+} from '@common/utils';
+import { HandshakeKeyPair, ReduxIPC } from '@types';
+import { safeJSONParse } from '@utils';
+
+interface HandshakeState {
+  publicKey?: string;
+  privateKey?: string;
+
+  isHandshaken: boolean;
+
+  targetPublicKey?: string;
+}
+
+const initialState: HandshakeState = {
+  isHandshaken: false
+};
+
+const sliceName = 'handshake';
+
+const slice = createSlice({
+  name: sliceName,
+  initialState,
+  reducers: {
+    createKeyPair(_, __: PayloadAction<boolean | undefined>) {
+      // noop
+    },
+    sendPublicKey(_, __: PayloadAction<string>) {
+      // noop
+    },
+    setHandshaken(state, action: PayloadAction<boolean>) {
+      state.isHandshaken = action.payload;
+    },
+    setKeyPair(state, action: PayloadAction<HandshakeKeyPair>) {
+      state.publicKey = action.payload.publicKey;
+      state.privateKey = action.payload.privateKey;
+    },
+    setTargetPublicKey(state, action: PayloadAction<string>) {
+      state.targetPublicKey = action.payload;
+    }
+  }
+});
+
+export const {
+  createKeyPair,
+  setKeyPair,
+  sendPublicKey,
+  setHandshaken,
+  setTargetPublicKey
+} = slice.actions;
+
+export default slice;
+
+export const getHandshakeState = createSelector(
+  (state: { handshake: HandshakeState }) => state,
+  (state) => state.handshake
+);
+
+export const getPublicKey = createSelector(getHandshakeState, (state) => state.publicKey);
+export const getPrivateKey = createSelector(getHandshakeState, (state) => state.privateKey);
+export const getHandshaken = createSelector(getHandshakeState, (state) => state.isHandshaken);
+
+export const getTargetPublicKey = createSelector(
+  getHandshakeState,
+  (state) => state.targetPublicKey
+);
+
+export const postHandshake = createAction(`${sliceName}/postHandshake`);
+
+export function* handshakeSaga(ipc: ReduxIPC) {
+  yield all([
+    ipcWorker(ipc),
+    takeLatest(createKeyPair.type, createKeyPairWorker),
+    takeEvery(sendPublicKey.type, setPublicKeyWorker)
+  ]);
+}
+
+export const subscribe = (ipc: ReduxIPC) => {
+  return eventChannel((emitter) => {
+    const unsubcribe = ipc.on((_: Event, action: string) => {
+      if (action) {
+        emitter(action);
+      }
+    });
+
+    return () => {
+      unsubcribe();
+    };
+  });
+};
+
+export function* putJson(json: string, allowInsecure: boolean = false): SagaIterator {
```

Comment:
Can we just call `allowInsecure` `isDecrypted` or something like that? May be clearer